File: Data/Complex.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace MD.Data
{
    /// <summary>
    /// Represents a double-precision complex number.
    /// </summary>
    public struct Complex
    {
        public Complex(double Real, double Imag)
        {
            this.Real = Real;
            this.Imag = Imag;
        }

        public Complex(double Real)
        {
            this.Real = Real;
            this.Imag = 0.0;
        }

        public override string ToString()
        {
            return this.Real.ToString() + " + " + this.Imag.ToString() + "i";
        }

        /// <summary>
        /// Gets the value of e to the power of this complex number.
        /// </summary>
        public Complex Exp
        {
            get
            {
                double rexp = Math.Exp(this.Real);
                double icos = Math.Cos(this.Imag);
                double isin = Math.Sin(this.Imag);
                return new Complex(rexp * icos, rexp * isin);
            }
        }

        /// <summary>
        /// Gets the argument of the complex number.
        /// </summary>
        public double Arg
        {
            get
            {
                return Math.Atan2(this.Imag, this.Real);
            }
        }

        /// <summary>
        /// Get the square of the absolute value of this complex number.
        /// </summary>
        public double SquareAbs
        {
            get
            {
                return this.Real * this.Real + this.Imag * this.Imag;
            }
        }

        /// <summary>
        /// Gets the absolute value of this complex number.
        /// </summary>
        public double Abs
        {
            get
            {
                return Math.Sqrt(this.SquareAbs);
            }
        }

        /// <summary>
        /// Gets the value of this complex number when multiplied by I.
        /// </summary>
        public Complex TimesI
        {
            get
            {
                return new Complex(-this.Imag, this.Real);
            }
        }

        public static Complex operator +(double A, Complex B)
        {
            return new Complex(A + B.Real, B.Imag);
        }

        public static Complex operator +(Complex A, double B)
        {
            return new Complex(A.Real + B, A.Imag);
        }

        public static Complex operator +(Complex A, Complex B)
        {
            return new Complex(A.Real + B.Real, A.Imag + B.Imag);
        }

        public static Complex operator -(double A, Complex B)
        {
            return new Complex(A - B.Real, -B.Imag);
        }

        public static Complex operator -(Complex A, double B)
        {
            return new Complex(A.Real - B, A.Imag);
        }

        public static Complex operator -(Complex A, Complex B)
        {
            return new Complex(A.Real - B.Real, A.Imag - B.Imag);
        }

        public static Complex operator *(double A, Complex B)
        {
            return new Complex(A * B.Real, A * B.Imag);
        }

        public static Complex operator *(Complex A, double B)
        {
            return new Complex(A.Real * B, A.Imag * B);
        }

        public static Complex operator *(Complex A, Complex B)
        {
            return new Complex(A.Real * B.Real - A.Imag * B.Imag, A.Imag * B.Real + A.Real * B.Imag);
        }

        public static Complex operator /(double A, Complex B)
        {
            return new Complex(A * B.Real, -A * B.Imag) / B.SquareAbs;
        }

        public static Complex operator /(Complex A, double B)
        {
            return new Complex(A.Real / B, A.Imag / B);
        }

        public static Complex operator /(Complex A, Complex B)
        {
            return new Complex(A.Real * B.Real + A.Imag * B.Imag, A.Imag * B.Real - A.Real * B.Imag) / B.SquareAbs;
        }

        public static implicit operator Complex(double A)
        {
            return new Complex(A);
        }

        /// <summary>
        /// The real part of the complex number.
        /// </summary>
        public double Real;

        /// <summary>
        /// The imaginary part of the complex number.
        /// </summary>
        public double Imag;
    }

}

File: Source/Complex.fs
namespace MD

open System
open System.Runtime.InteropServices

/// Represents a complex number.
[<StructLayout (LayoutKind.Sequential)>]
type Complex (real : float, imag : float) =
    struct

        /// Gets zero as a complex number.
        static member Zero = new Complex (0.0, 0.0)

        /// Adds two complex numbers.
        static member (+) (a : Complex, b : Complex) = new Complex (a.Real + b.Real, a.Imag + b.Imag)
        
        /// Adds a complex number to a real.
        static member (+) (a : Complex, b : float) = new Complex (a.Real + b, a.Imag)

        /// Subtracts two complex numbers.
        static member (-) (a : Complex, b : Complex) = new Complex (a.Real - b.Real, a.Imag - b.Imag)

        /// Subtracts a real from a complex number.
        static member (-) (a : Complex, b : float) = new Complex (a.Real - b, a.Imag)

        /// Multiples two complex numbers.
        static member (*) (a : Complex, b : Complex) = 
            new Complex (a.Real * b.Real - a.Imag * b.Imag, a.Imag * b.Real + a.Real * b.Imag)

        /// Multiples a complex number by a real.
        static member (*) (a : Complex, b : float) = 
            new Complex (a.Real * b, a.Imag * b)

        /// Divides two complex numbers.
        static member (/) (a : Complex, b : Complex) = 
            let iden = 1.0 / (b.Real * b.Real + b.Imag * b.Imag)
            let real = (a.Real * b.Real + a.Imag * b.Imag) * iden
            let imag = (a.Imag * b.Real - a.Real * b.Imag) * iden
            new Complex (real, imag)

        /// Gets e to the power of a number multiplied by i.
        static member ExpImag (imag : float) =
            new Complex (cos imag, -(sin imag))

        /// Gets the real component of this number.
        member this.Real = real

        /// Gets the imaginary component of this number.
        member this.Imag = imag

        /// Gets the absolute value, or magnitude of this number.
        member this.Abs = sqrt (real * real + imag * imag)

        /// Gets the square of the absolute value of this number.
        member this.SquareAbs = real * real + imag * imag

        /// Gets the phase of this complex number in radians
        member this.Phase = atan2 imag real

        /// Multiplies this number by i.
        member this.TimesI = new Complex (-imag, real)

        /// Gets the complex conjugate of this number.
        member this.Conjugate = new Complex (real, -imag)

        override this.ToString() = 
            if imag >= 0.0 then String.Format("{0} + {1}i", real, imag)
            else String.Format("{0} - {1}i", real, -imag)
    end